examples/roitberg/roitberg.py

```diff
@@ -15,71 +21,118 @@ def convert_species_to_atomic_nums(s):
+  return os.path.abspath(os.path.expanduser(os.path.join(*dir_segments)))
+
+def path_not_empty(d):
+  d = path(d)
+  return os.path.exists(d) and len(os.listdir(d)) > 1
+
+def find_or_create_path(*dir_segments):
+  """
+  Pretty much what it sounds like: finds a directory path or creates one if it
+  doesn't exist.
+
+  Also adds in the functionality of os.path.join, for extra helpfulness: the
+  dir_segments will be joined together, then checked for existence.
+  """
+  d = path(*dir_segments)
+
+  if not os.path.exists(d):
+    os.makedirs(d, exist_ok=True)
+
+  return d
+
+def setup_work_dirs(args):
+  """
+  Sets up the work directory structure beneath the directory in +args.work_dir+.
+
+  If args.work_dir is not a valid directory, errors out with a warning to the user,
+  on the theory that the program shouldn't be chucking lots of data in places the
+  user isn't explicitly aware of.
+
+  Below args.work_dir, creates new subdirectories with impunity. The various work
+  directories will be set in +args+ for later reference.
+
+  Parameters:
+    args - the result of calling ArgumentParser.parse_args()
+  """
+  args.work_dir = path(args.work_dir)
+
+  if not os.path.exists(args.work_dir):
+    # we don't create the work dir if it doesn't already exist:
+    # don't want to silently dump big files on the user's FS without
+    # their knowledge.
+    sys.stderr.write("Work directory '%s' does not exist. "
+                     "Create it, or specify a new directory with the -w option.\n" % args.work_dir)
+    sys.exit(-1)
+  elif args.clean_work_dir:
+    # python doesn't have an easy way to remove all files in a directory
+    # this will *probably* work, but might fail on weird setups.
+    shutil.rmtree(args.work_dir)
+    os.makedirs(args.work_dir, exist_ok=True)
+
+  args.data_dir = find_or_create_path(args.work_dir, "datasets")
+  args.model_dir = find_or_create_path(args.work_dir, "models")
+
+  args.all_dir = find_or_create_path(args.data_dir, "all")
+  args.test_dir = find_or_create_path(args.data_dir, "test")
+  args.fold_dir = find_or_create_path(args.data_dir, "fold")
+
+  args.train_dir = find_or_create_path(args.fold_dir, "train")
+  args.temp_dir = find_or_create_path(args.fold_dir, "temp")
+  args.valid_dir = find_or_create_path(args.fold_dir, "valid")
+
+  return args
+
+
+def find_training_data(base_dir, max_gdb_level):
+  """
+  Generates filenames for GDB files and checks for their existence.
+  Fails out with user error if any file is not found.
 
-# replace with your own scratch directory
-data_dir = "/media/yutong/fast_datablob/datasets"
-model_dir = "/media/yutong/fast_datablob/models"
-# feat_train_dir = "/media/yutong/fast_datablob/feat"
-# feat_valid_dir = "/media/yutong/fast_datablob/valid"
+  Returns:
+    list of validated file path strings.
+  """
+  base_dir = os.path.abspath(os.path.expanduser(base_dir))
+  files = [os.path.join(base_dir, "ani_gdb_s%02d.h5" % i) for i in range(max_gdb_level, 0, -1)]
```

Comment:
Is there any reason why we do this backwards (as opposed to range (0,9))?

Reply:
Because I was just blindly regenerating the array in the old code, lol. I'll reverse it.